### PR TITLE
Implemented pallet for cross-chain governance on Solo-chain

### PR DESCRIPTION
**Pull request by zktony**
## Describe your changes
Thea Governance Pallet on Solo-chain.
The following things are implemented:-
* Apply for Candidature for Thea Council. (Provide keys for current network)
* Approve candidature. (Need majority of General Council or Root)
* Remove existing members. (Need majority of General Council or Root)

## Issue ticket number and link
Partially closes https://github.com/Polkadex-Substrate/thea/issues/78

### CODE REVIEW

**Comment by serhii-temchenko** [comment]
@zktony is it ready to be reviewed? If so - please add a label so I could see that it is ready.

**Comment by zktony** [comment]
> @zktony is it ready to be reviewed? If so - please add a label so I could see that it is ready.

Sorry, I was waiting for checks to pass.

File: pallets/thea-cross-chain-governance/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	/// Configure the pallet by specifying the parameters and types on which it depends.
+	#[pallet::config]
+	pub trait Config: frame_system::Config + pallet_balances::Config {
+		/// Because this pallet emits events, it depends on the runtime's definition of an event.
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+		/// Stake required to apply for candidature
+		type StakingAmount: Get<u128>;
```
Comment (Gauthamastro) [inline pallets/thea-cross-chain-governance/src/lib.rs]: Use the balance type from primitives

File: pallets/thea-cross-chain-governance/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+	}
+
+	// Dispatchable functions allows users to interact with the pallet and invoke state changes.
+	// These functions materialize as "extrinsics", which are often compared to transactions.
+	// Dispatchable functions must be annotated with a weight and must return a DispatchResult.
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Apply for candidature
+		#[pallet::weight(10_000 + T::DbWeight::get().writes(1))]
+		pub fn apply_for_candidature(origin: OriginFor<T>, keys_list: KeysMap) -> DispatchResult {
```
Comment (Gauthamastro) [inline pallets/thea-cross-chain-governance/src/lib.rs:97]: Also, add the check if an identity is attached to this account, if not reject

File: pallets/thea-cross-chain-governance/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+		type StakingReserveIdentifier: Get<<Self as pallet_balances::Config>::ReserveIdentifier>;
+		/// CouncilHandlerOrigin
+		type CouncilHandlerOrigin: EnsureOrigin<<Self as frame_system::Config>::Origin>;
+	}
+
+	#[pallet::storage]
+	#[pallet::getter(fn active_members)]
+	/// Currently active thea council member
+	pub(super) type ActiveMembers<T: Config> =
+		StorageMap<_, Blake2_128Concat, T::AccountId, KeysMap, OptionQuery>;
```
Comment (Gauthamastro) [inline pallets/thea-cross-chain-governance/src/lib.rs:46]: I prefer it to be a boundedVec than storagemap

File: pallets/thea-cross-chain-governance/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+
+	#[pallet::storage]
+	#[pallet::getter(fn candidates)]
+	/// New candidates
+	pub(super) type Candidates<T: Config> =
+		StorageMap<_, Blake2_128Concat, T::AccountId, KeysMap, OptionQuery>;
```
Comment (Gauthamastro) [inline pallets/thea-cross-chain-governance/src/lib.rs:52]: Same here.

File: pallets/thea-cross-chain-governance/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+			Self::do_apply(&who, keys_list)?;
+			Self::deposit_event(Event::<T>::NewAccountAdded(who));
+			Ok(())
+		}
+
+		/// Approve candidate request
+		#[pallet::weight(10_000 + T::DbWeight::get().writes(1))]
+		pub fn approve_candidature(
+			origin: OriginFor<T>,
+			candidate: T::AccountId,
```
Comment (Gauthamastro) [inline pallets/thea-cross-chain-governance/src/lib.rs]: A bounded vector of accounts should be provided.

**Review by Gauthamastro — CHANGES_REQUESTED** [review]

File: pallets/thea-cross-chain-governance/Cargo.toml [hunk 6]
```diff
@@ -0,0 +1,40 @@
+[package]
+name = "thea-cross-chain-governance"
+version = "0.1.0"
+edition = "2021"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+parity-scale-codec = {workspace = true, default-features = false }
```
Comment (serhii-temchenko) [inline pallets/thea-cross-chain-governance/Cargo.toml]: Please add space after `{`

File: pallets/thea-cross-chain-governance/Cargo.toml [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+log = { workspace = true, default-features = false }
+scale-info =  { workspace = true, default-features = false, features = ["derive"] }
+frame-benchmarking = { default-features = false, workspace = true , optional = true}
```
Comment (serhii-temchenko) [inline pallets/thea-cross-chain-governance/Cargo.toml]: Please add space before `}`

File: pallets/thea-cross-chain-governance/src/lib.rs (hunk 2 above)
Comment (serhii-temchenko) [inline pallets/thea-cross-chain-governance/src/lib.rs:97]: Please add parameters in to the dockblock with proper description.

File: pallets/thea-cross-chain-governance/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+		) -> DispatchResult {
```
Comment (serhii-temchenko) [inline pallets/thea-cross-chain-governance/src/lib.rs]: Please add parameters in to the dockblock with proper description.

File: pallets/thea-cross-chain-governance/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+			T::CouncilHandlerOrigin::ensure_origin(origin)?;
+			Self::do_approve(&candidate)?;
+			Self::deposit_event(Event::<T>::CandidateApproved(candidate));
+			Ok(())
+		}
+
+		/// Add keys for new Networks
+		#[pallet::weight(10_000 + T::DbWeight::get().writes(1))]
+		pub fn add_new_keys(origin: OriginFor<T>, new_keys: KeysMap) -> DispatchResult {
```
Comment (serhii-temchenko) [inline pallets/thea-cross-chain-governance/src/lib.rs:130]: Please add parameters in to the dockblock with proper description.

File: pallets/thea-cross-chain-governance/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+			let member = ensure_signed(origin)?;
+			Self::do_add_keys(&member, new_keys)?;
+			Self::deposit_event(Event::<T>::NewKeysAdded(member));
+			Ok(())
+		}
+
+		/// Remove from Active List
+		#[pallet::weight(10_000 + T::DbWeight::get().writes(1))]
+		pub fn remove_candidate(origin: OriginFor<T>, candidate: T::AccountId) -> DispatchResult {
```
Comment (serhii-temchenko) [inline pallets/thea-cross-chain-governance/src/lib.rs]: Please add parameters in to the dockblock with proper description.

File: runtime/Cargo.toml [hunk 11]
```diff
@@ -100,6 +100,7 @@ polkadex-primitives = { workspace = true, default-features = false }
 pdex-migration = { path = "../pallets/pdex-migration", default-features = false }
 asset-handler = { path = "../pallets/asset-handler", default-features = false }
 chainbridge = { path ="../pallets/chainbridge", default-features = false }
+thea-cross-chain-governance = {path = "../pallets/thea-cross-chain-governance", default-features = false}
```
Comment (serhii-temchenko) [inline runtime/Cargo.toml]: Same with spaces before and after `{}`

**Review by serhii-temchenko — CHANGES_REQUESTED** [review]
Minor improvements required.
Thanks for really nice written tests!

One more question - do we need benchmarks here? If so - please add task for it.

File: pallets/thea-cross-chain-governance/src/lib.rs (hunk 1 above)
Comment (zktony) [reply pallets/thea-cross-chain-governance/src/lib.rs]: done

File: runtime/Cargo.toml (hunk 11 above)
Comment (zktony) [reply runtime/Cargo.toml]: done

**Comment by serhii-temchenko** [comment]
@zktony I'll move this PR to `InProgress` label since it looks like work on comments still ongoing. Please move it back to `PleaseReview` when it will be ready so I could understand that it could be re-reviewed.

**Review by serhii-temchenko — APPROVED** [review]

**Review by Gauthamastro — APPROVED** [review]

**Pull request merged**